[MRG+1] Remove check_array for X in DummyRegressor/DummyClassifier and replace X.shape[0] with _num_samples

vrishank97 opened the pull request:
Fixes #9832

Adds dtype="None" and ensure_min_features=0 as params for check_array in DummyRegressor predict.

vrishank97 committed:
Add params for check_array in DummyRegressor predict

vrishank97 committed:
Add params for check_array in DummyRegressor predict

jnothman reviewed (commented):
This needs a test.

Perhaps just discard the check_array and use `utils.validation._num_samples`. Note sure if this will pass the common tests, though.

vrishank97 committed:
Fixed typo in check_array params

vrishank97 committed:
Fixed typo in check_array params

vrishank97 commented:
I haven't made tests before. Would adding the code to reproduce the bug from the report to sklearn/tests/test_dummy.py in the form of function suffice?

jnothman commented:
ideally you would do that, but with a numpy array rather than a pandas
DataFrame.

vrishank97 committed:
Add test for object val for DummyRegressor

vrishank97 committed:
Add test for object val for DummyRegressor

vrishank97 committed:
Remove whitespace and new line

vrishank97 committed:
Remove whitespace and new line

vrishank97 committed:
removed whitespace

vrishank97 committed:
removed whitespace

jnothman commented on sklearn/tests/test_dummy.py:
```diff
@@ -617,3 +617,14 @@ def test_dummy_regressor_on_nan_value():
+
+   
+def test_dummy_regressor_on_object_value():
+    X = np.array(['foo', 'bar', 'baz'])
```
this won't actually have dtype object (check for yourself) so the test is incorrectly named.

vrishank97 replied:
Sorry. I'll change it to string.

vrishank97 committed:
Removed extra whitespace

vrishank97 committed:
Removed extra whitespace

vrishank97 committed:
change test name

Changed test name from test_dummy_regressor_on_object_value to test_dummy_regressor_on_string_values.

vrishank97 committed:
change test name

Changed test name from test_dummy_regressor_on_object_value to test_dummy_regressor_on_string_values.

vrishank97 commented:
I have changed the test name. @jnothman is there anything else that needs to be changed?

jnothman reviewed (approved):
LGTM!

qinhanmin2014 commented:
ping @vrishank97 Some personal concerns:
(1) Why is the check only modified in predict, not in fit? fit and perdict do the same check here.
(2) Seems that the test will pass even in current master. E.g.
```python
X = np.array(['a', 'b', 'c']).reshape(-1, 1)
check_array(X, dtype='numeric') # no error
```
(3) Seems that the original code snippet in the issue will still get an error.

qinhanmin2014 commented:
So seems that there's something wrong with check_array (#10229).
@vrishank97 Could you please consider to copy the change to fit and predict_preba (and maybe DummyRegressor) and extend the test accordingly? Thanks.

qinhanmin2014 reviewed (commented):
ping @vrishank97 Could you please take some time to finish it. Otherwise I'll pick it up.
Even with current check_array, I believe we can still construct a regression test using astype("O").

amueller commented:
We should probably check with a zero feature array as well, right?

vrishank97 committed:
Modify params for check_array in method fit

vrishank97 committed:
Modify params for check_array in method fit

vrishank97 committed:
Add params to check_array in DummyClassifier

vrishank97 committed:
Add params to check_array in DummyClassifier

qinhanmin2014 commented:
@vrishank97 
There's predict_proba in DummyClassifier.
For the test, I think you need to cover all your changes. Also make sure that your test fail on master but success in the PR.

vrishank97 committed:
Modify check_array params in predict_proba for DummyClassifier

vrishank97 committed:
Modify check_array params in predict_proba for DummyClassifier

vrishank97 committed:
Modified test

vrishank97 committed:
Modified test

qinhanmin2014 commented on sklearn/tests/test_dummy.py:
```diff
@@ -617,3 +617,15 @@ def test_dummy_regressor_on_nan_value():
+    y_pred = cls.predict(X)
+    assert_array_equal(y_pred, y_expected)
+
+test_dummy_regressor_on_string_values()
```
test dummy classifier?

vrishank97 replied:
Just made the test. Pushing now

vrishank97 committed:
Modify DummyClassifier test for object values

vrishank97 committed:
Modify DummyClassifier test for object values

vrishank97 committed:
Add test for predict_proba for DummyClassifier

vrishank97 committed:
Add test for predict_proba for DummyClassifier

vrishank97 commented:
@qinhanmin2014 
I've added tests for predict and predict_proba methods for DummyClassifier. Can you have a look at it now?

qinhanmin2014 commented on sklearn/dummy.py:
```diff
@@ -107,7 +107,8 @@ def fit(self, X, y, sample_weight=None):
         X = check_array(X, accept_sparse=['csr', 'csc', 'coo'],
-                        force_all_finite=False)
+                        force_all_finite=False, dtype=None,
+                        ensure_min_features=0)
```
Is there any specific reason adding ensure_min_features=0? No other place in current codebase do so. It's a classifier/regressor so I suppose we should at least feed it with some features?

vrishank97 replied:
The proposal in #9832 suggested than we make make the input validation less strict by adding ensure_min_features=0. However, I agree with you that since it is a classifier/regressor we should feed it some features.

qinhanmin2014 replied:
OK, seems that jnothman and amueller don't oppose. I won't oppose although it seems a bit strange. Please add similar test for a zero feature array. I think my LGTM is close :)

vrishank97 replied:
I'm not sure what a zero feature array is. Is it an array of null values or an array of zeros?

qinhanmin2014 replied:
Something like `np.array([[],[],[]])` (this is why I think it's strange.). It's always good to make sure that your test fail on master and pass in the PR.

vrishank97 committed:
Fixed style issues

vrishank97 committed:
Fixed style issues

vrishank97 committed:
Add test for zero feature array for DummyRegressor

vrishank97 committed:
Add test for zero feature array for DummyRegressor

vrishank97 committed:
Add test for zero feature arrays for DummyClassifier

vrishank97 committed:
Add test for zero feature arrays for DummyClassifier

vrishank97 committed:
Remove typo

vrishank97 committed:
Remove typo

vrishank97 committed:
Removed typo

vrishank97 committed:
Removed typo

qinhanmin2014 commented on sklearn/tests/test_dummy.py:
```diff
@@ -617,3 +617,46 @@ def test_dummy_regressor_on_nan_value():
+    assert_array_equal(y_pred_proba, y_proba_expected)
+
+
+def test_dummy_regressor_on_zero_arrays():
```
zero_feature_array

qinhanmin2014 commented on sklearn/tests/test_dummy.py:
```diff
@@ -617,3 +617,46 @@ def test_dummy_regressor_on_nan_value():
+    X = np.array([[], [], []])
+    y = np.array([2, 2, 2])
+    y_expected = np.array([2, 2, 2])
+    cls = DummyRegressor(strategy='mean')
```
remove strategy='mean'?

qinhanmin2014 commented on sklearn/tests/test_dummy.py:
```diff
@@ -617,3 +617,46 @@ def test_dummy_regressor_on_nan_value():
+    y_expected = np.array([2, 2, 2])
+    cls = DummyClassifier()
+    cls.fit(X, y)
+    y_pred = cls.predict(X)
```
predict_proba?

vrishank97 replied:
Oops. I've pushed the changes.

qinhanmin2014 reviewed (commented):
It might be better to use the same data for the four tests.

vrishank97 committed:
Add test for predict_proba

vrishank97 committed:
Add test for predict_proba

vrishank97 committed:
Common data for new tests

vrishank97 committed:
Common data for new tests

vrishank97 committed:
Remove typos

vrishank97 committed:
Remove typos

qinhanmin2014 commented on sklearn/tests/test_dummy.py:
```diff
@@ -617,3 +617,49 @@ def test_dummy_regressor_on_nan_value():
+
+
+def test_dummy_regressor_on_object_values():
+    X = np.array([[], [], []], dtype=object)
```
too much copy-paste :)

qinhanmin2014 replied:
Maybe not using zero feature array here is better. I prefer the previous version.

qinhanmin2014 commented on sklearn/tests/test_dummy.py:
```diff
@@ -617,3 +617,49 @@ def test_dummy_regressor_on_nan_value():
+    assert_array_equal(y_pred_proba, y_proba_expected)
+
+
+def test_dummy_regressor_on_zero_feature_arrays():
```
zero_feature_array

vrishank97 replied:
Oops. Done

vrishank97 committed:
Removed more typos

vrishank97 committed:
Removed more typos

vrishank97 committed:
Remove zero feature arrays from tests on object values

vrishank97 committed:
Remove zero feature arrays from tests on object values

vrishank97 committed:
Remove typo

vrishank97 committed:
Remove typo

qinhanmin2014 reviewed (approved):
LGTM ping @jnothman 
Latest change since your +1:
(1) extend the change to all similar checks in DummyRegressor & DummyClassifier
(2) use object array instead of string array to ensure that the test is regression under current check_array
(3) add test for zero feature array

jnothman reviewed (commented):
Thanks. Before I merge: Can you remind me why we want to check X at all here, and not just use `sklearn.utils.validation._num_samples`

qinhanmin2014 commented:
@jnothman I think check_array still has something to check in this case although we have removed some restrictions, e.g., I suppose we want a 2d array to serve as training vectors, which can't be done by _num_samples.

jnothman commented:
But as far as I can tell, we only ever use X for its length here. I reckon
it should be possible to pass in any X that is used for any other
classifier within reason, i.e. it should have the same length as y, but we
need not waste effort converting it to an array here... WDYT?

On 14 December 2017 at 12:25, Hanmin Qin <notifications@github.com> wrote:

> @jnothman <https://github.com/jnothman> I think check_array still has
> something to check in this case although we have removed some restrictions,
> e.g., I suppose we want a 2d array to serve as training vectors, which
> can't be done by _num_samples.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9835#issuecomment-351579528>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6wcj1AbyljlGOrVaofoUQKtDVvqiks5tAHkhgaJpZM4PlT4q>
> .
>

qinhanmin2014 commented:
@jnothman Indeed, seems that we only use X.shape[0] in DummyRegressor/DummyClassifier. But I might think a check_array for X is still necessary. E.g., I think it will be good to raise an error if users pass in something other than 2d array (like [1, 2, 3]). Also, I suppose a check_array here won't cost too much.
Btw, we are missing check_consistent_length(X, y, sample_weight) in DummyClassifier

vrishank97 commented:
Should I put in a check_consistent_length in DummyClassifier?

jnothman commented:
Despite previously approving, I think I'm sticking with the idea that we
should just be using _num_samples. I just don't see the point in knocking
someone back for wanting to compare a dummy estimator to some custom
estimator that just takes a 1d array or some obscure data structure that
has .shape or __len__.

qinhanmin2014 commented:
@jnothman But we still need X.shape[0] during the calculation, I suppose _num_samples can't ensure that?

qinhanmin2014 commented:
@jnothman So you mean using _num_samples instead of X.shape[0] and remove check_array for X?

jnothman commented:
precisely

qinhanmin2014 commented:
+1 ping @vrishank97 Sorry for changing the solution. Could you please take some time to finish it?
You also need to add check_consistent_length for DummyClassifier and rewrite the test.

vrishank97 committed:
Remove check_array for X

vrishank97 committed:
Remove check_array for X

vrishank97 commented:
Sure. No problem

vrishank97 commented:
@qinhanmin2014 Do we really need to rewrite the tests? I just ran them and they fail on master but work with the PR.

jnothman commented:
well you could update the tests to ensure a list of strings, or a list of
sets, or a 3d array that would have failed check_array now works. You can
also update the parameter description to say it only requires a length.

qinhanmin2014 commented:
@vrishank97 Personally suggestions for your consideration :)

> Do we really need to rewrite the tests? I just ran them and they fail on master but work with the PR.

It is the case but I suppose it might be better to simplify the test, especially for such minor issue. I think the last 6 tests can all be removed (including the two checking for nan) because we no longer use check_array. I might prefer to construct two tests (one for classifier, another for regressor), pick some inputs which are not accepted previously (e.g., the instructions from jnothman above) and check the result.

vrishank97 committed:
Add test on 3D array for DummyRegressor and DummyClassifier

vrishank97 committed:
Add test on 3D array for DummyRegressor and DummyClassifier

vrishank97 commented:
I have replaces the previous tests with a test on 3D array for DummyClassifier and DummyRegressor

qinhanmin2014 commented:
@vrishank97 You also need to modify the doc. Please find all the explanations for parameter X and modify them accordingly.

vrishank97 committed:
Update parameter description for X

vrishank97 committed:
Update parameter description for X

vrishank97 commented:
Updated the parameter descriptions.

qinhanmin2014 commented:
Maybe need to state something like length = n_samples to be consistent with the description of y below.

vrishank97 committed:
Add length = n_samples for X

vrishank97 committed:
Add length = n_samples for X

vnavkal committed:
DOC fix typo in estimator documentation (#10344)

lesteve committed:
MAINT make build_doc.sh work if _VERSION env variables are not set

amueller committed:
FIX cython cluster module name error (#10343)

lesteve committed:
FIX CircleCI deploy command.

Bash regexes should not be quoted.

vrishank97 commented:
Does this need any other improvements?

qinhanmin2014 reviewed (approved):
LGTM from my side ping @jnothman

jnothman commented on sklearn/dummy.py:
```diff
@@ -91,9 +92,8 @@ def fit(self, X, y, sample_weight=None):
-            Training vectors, where n_samples is the number of samples
-            and n_features is the number of features.
+        X : {array-like, sparse matrix}
+            Input vector, requires length = n_samples
```
Hmm... This doesn't make sense. It's not ordinarily a vector but a matrix. And it can't require something that's defined due to it. I think all I mean is that `{array-like, sparse matrix}` can be extended to `{array-like, object with finite length or shape}`

jnothman reviewed (commented):
Please also add a what's new entry under API changes

vrishank97 committed:
Update dummy.py

vrishank97 committed:
Update dummy.py

qinhanmin2014 commented:
@vrishank97 Please also update what's new accordingly. Also, I think we might keep length = n_samples.

vrishank97 committed:
Update dummy.py

vrishank97 committed:
Update dummy.py

qinhanmin2014 commented:
@vrishank97 Please also update what's new(docs/whats_new/v0.20.rst) accordingly.

vrishank97 committed:
Update what's new

vrishank97 committed:
Merge branch 'patch-3' of https://github.com/vrishank97/scikit-learn into patch-3

vrishank97 commented:
I've updated the what's new section.

qinhanmin2014 commented on doc/whats_new/v0.20.rst:
```diff
@@ -116,6 +116,10 @@ Classifiers and regressors
 
+- :class:'dummy.DummyClassifier' and :class:'dummy.DummyRegresssor' now
+  only requires X to be an object with finite length or shape.
+  :issue:'9832' by :user:'Vrishank Bhardwaj <vrishank97>'.
```
You should use ` instead of ', please follow what others' are doing and check the result from Circle CI.

vrishank97 replied:
Oops.

qinhanmin2014 reviewed (commented):
@vrishank97 Please avoid making unrelated changes.

vrishank97 committed:
Remove typos

vrishank97 commented:
I've removed the typos. I think I also accidentally included a few unrelated commits by others while rebasing. How should I remove them?

qinhanmin2014 commented:
At least you might recover the code according to the diff.

vrishank97 committed:
remove diff

vrishank97 committed:
remove diff

vrishank97 committed:
remove diff

vrishank97 committed:
remove diff

vrishank97 commented:
Removed the diffs.

qinhanmin2014 reviewed (approved):
LGTM ping @jnothman

lesteve commented:
Wow I have no idea about the root reason for this ... but it looks like merging this one reverted some changes in master, probably because of the weird history ...

* the diff in this PR changes 3 file
* the merge commit https://github.com/scikit-learn/scikit-learn/commit/58ff9b8600777166a234f96e418834d3328d1cea changes 7 files, reverting some changes in master ...

I'll fix that in master but this is something to keep in mind (check the history has no quirks) before doing squash+merge.

lesteve commented:
Reverted the unwanted changes in master in e824fa1

jnothman commented:
How strange, thanks for cleaning up after me!

qinhanmin2014 commented:
Thanks @lesteve Really sorry about that. Also curious about the reason.

lesteve commented:
> Thanks @lesteve Really sorry about that. Also curious about the reason.

No worries! Anyone with a half-baked explanation I am interested ...

jnothman commented:
Perhaps it's worth notifying support@github.com that there is a bug in squash and merge...?

lesteve commented:
Maybe, before classifying it as a bug, it would be nice if someone can do a bit of research about
- what git commands are executed in the background for the squash and merge
- whether similar behaviours have been reported in the past
